pkg/shutdown/shutdown.go

```diff
@@ -0,0 +1,1030 @@
+
+	// TimeoutForceShutdownWindow - Time window for second signal to trigger force shutdown
+	// Users can send the same signal again within this window to bypass graceful shutdown
+	TimeoutForceShutdownWindow = 5 * time.Second
+)
+
+// Test timeout constants for consistent test behavior
+const (
+	// TimeoutTestVeryFast - For very fast test operations and quick timeouts
+	TimeoutTestVeryFast = 50 * time.Millisecond
+
+	// TimeoutTestFast - For fast test components and quick assertions
+	TimeoutTestFast = 100 * time.Millisecond
+
+	// TimeoutTestStandard - Standard test timeout for component verification
+	TimeoutTestStandard = 200 * time.Millisecond
+)
+
+// ShutdownState represents the various states of shutdown
+type ShutdownState int
+
+const (
+	StateIdle ShutdownState = iota
+	StateInitiated
+	StateInProgress
+	StateCompleted
+	StateFailed
+)
+
+func (s ShutdownState) String() string {
+	switch s {
+	case StateIdle:
+		return "idle"
+	case StateInitiated:
+		return "initiated"
+	case StateInProgress:
+		return "in_progress"
+	case StateCompleted:
+		return "completed"
+	case StateFailed:
+		return "failed"
+	default:
+		return "unknown"
+	}
+}
+
+// ComponentStatus tracks the status of an individual component during shutdown
+type ComponentStatus struct {
+	Name      string
+	Status    string // "pending", "in_progress", "success", "error", "timeout"
+	StartTime *time.Time
+	EndTime   *time.Time
+	Duration  time.Duration
+	Error     string
+}
+
+// ShutdownStatus represents the current shutdown state of the service for API responses
+type ShutdownStatus struct {
+	IsShuttingDown      bool                 `json:"isShuttingDown"`
+	ShutdownInitiated   *time.Time           `json:"shutdownInitiated,omitempty"`
+	ActiveComponents    []string             `json:"activeComponents,omitempty"`
+	CompletedComponents []CompletedComponent `json:"completedComponents,omitempty"`
+	State               string               `json:"state"` // "idle", "initiated", "in_progress", "completed", "failed"
+	Message             string               `json:"message,omitempty"`
+}
+
+// CompletedComponent represents a component that has finished shutting down
+type CompletedComponent struct {
+	Name     string        `json:"name"`
+	Status   string        `json:"status"` // "success", "error", "timeout"
+	Duration time.Duration `json:"duration"`
+	Error    string        `json:"error,omitempty"`
+}
+
+// ShutdownCallback is a function that performs cleanup for a component
+type ShutdownCallback func(ctx context.Context) error
+
+// Component represents a component that can be shut down gracefully
+type Component struct {
+	Name     string
+	Priority int           // Lower number = higher priority (shutdown first)
+	Timeout  time.Duration // Maximum time allowed for this component's shutdown
+	Callback ShutdownCallback
+}
+
+// MetricsCollector defines the interface for recording shutdown metrics
+type MetricsCollector interface {
```

Comment:
pretty sure we agreed in the design doc that we should use the existing metrics to "test" if the shutdown is doing what we need. this seems to be complexity that is not required.